Use vector of integer for per-thread bool indicators

suku248 opened the pull request:
- Resolve #1394 
- Rename  `class CompletedChecker` to `class PerThreadBoolIndicator` and adapt all instances accordingly
- Support subscript read/write access to elements of `PerThreadBoolIndicator`
- Use `std::vector` of `BoolIndicatorUInt64` instead of array of `bool` in `PerThreadBoolIndicator`
  (`BoolIndicatorUInt64` is a wrapper for `uint_fast64_t status_` that restricts functionality to `is_true/false()`, `set_true/false()`, and `logical_and()`)
- `SourceTable` members `is_cleared_` and `saved_entry_point_` are now of type `PerThreadBoolIndicator` instead of `std::vector<bool >`

heplesser commented on nestkernel/per_thread_bool_indicator.h:
```diff
@@ -0,0 +1,105 @@
+namespace nest
+{
+
+class BoolIndicatorUInt64
```
Could you add a brief doxygen comment in the class? I think the methods are so straightforward that they don't need further documentation.

suku248 replied:
Done.

heplesser commented on nestkernel/per_thread_bool_indicator.h:
```diff
@@ -0,0 +1,105 @@
+/**
+ * A thread-safe vector to keep track of the status across threads,
+ * for example during gather operations. Uses a vector of integers
+ * instead of a vector of bools to guarantee thread safety.
```
Maybe add a reference to issue #1394?

suku248 replied:
Done.

heplesser commented on nestkernel/source_table.cpp:
```diff
@@ -55,17 +55,15 @@ nest::SourceTable::initialize()
     sources_[ tid ].resize( 0 );
     resize_sources( tid );
-    is_cleared_[ tid ] = false;
-    saved_entry_point_[ tid ] = false;
```
Setting to false here is no longer needed because you now initialize properly on lines 48-49, isn't it? Just want to confirm my understanding of the code.

suku248 replied:
Correct!

heplesser commented on nestkernel/per_thread_bool_indicator.h:
```diff
@@ -0,0 +1,105 @@
+ * for example during gather operations. Uses a vector of integers
+ * instead of a vector of bools to guarantee thread safety.
+ */
+class PerThreadBoolIndicator
```
Do we need the `Bool` in the name? Wouldn't `PerThreadIndicator` suffice?

suku248 replied:
I would like to keep the `Bool` in there. We might want another more general `PerThreadIndicator` that can do more than `true` and `false` at some point.

heplesser commented on nestkernel/per_thread_bool_indicator.cpp:
```diff
@@ -0,0 +1,137 @@
+{
+}
+
+bool
```
Move to header file and `inline`, the same for `is_false()` and the `set_*()` methods.

suku248 replied:
Done.

suku248 commented on nestkernel/event_delivery_manager.cpp:
```diff
@@ -107,7 +107,6 @@ EventDeliveryManager::finalize()
   // clear the spike buffers
   std::vector< std::vector< std::vector< std::vector< Target > > > >().swap( spike_register_ );
   std::vector< std::vector< std::vector< std::vector< OffGridTarget > > > >().swap( off_grid_spike_register_ );
-  gather_completed_checker_.clear();
```
Previously, `CompletedChecker::clear()` was needed to deallocate the memory of the underlying C-style array and set the array pointer to `NULL`. As `PerThreadBoolIndicator` uses a `std::vector`, this is no longer required.

suku248 commented on nestkernel/connection_manager.cpp:
```diff
@@ -94,9 +94,9 @@ nest::ConnectionManager::initialize()
-  have_connections_changed_.resize( num_threads, true );
-  check_primary_connections_.resize( num_threads, false );
-  check_secondary_connections_.resize( num_threads, false );
+  have_connections_changed_.initialize( num_threads, true );
```
Well, actually it would make a lot more sense to initialize this to `false`, too.
I'll check if this breaks anything...

suku248 replied:
Travis indeed reported one failing Python test when `have_connections_changed_` is initialized to `false`. I did not see any problems when running `hpc_benchmark` with 32 MPI processes and 24 threads per process and none of the Python tests fails on my laptop. I guess it makes sense to look into this but I don't want to delay this PR.
@jakobj do you remember why `have_connections_changed_` needs to be initialized to `true`?

heplesser replied:
@stinebuu @hakonsbm Could you contribute to clarifying this?

stinebuu replied:
Unfortunately I don't know why it is initialized to `true`.

suku248 replied:
Let's not delay this PR because of this particular initialization.
I have issued another PR #1448 initializing `have_connections_changed_` to `false` such that we can continue the investigation there.

heplesser replied:
:+1:

suku248 commented on nestkernel/per_thread_bool_indicator.cpp:
```diff
@@ -0,0 +1,137 @@
+void
+BoolIndicatorUInt64::logical_and( const bool status )
+{
+  status_ = ( static_cast< bool >( status_ ) and status );
```
The right hand side results in a `bool`, which is then converted to either integer `0` or `1` when assigned to `status_`. This is consistent with the initial assignment of `true` and `false` to the constants `true_uint64` and `false_uint64`, respectively. We could use an `if/else` here to assign `true_uint64/false_uint64` but this might be a bit over the top.

suku248 commented on nestkernel/per_thread_bool_indicator.h:
```diff
@@ -0,0 +1,105 @@
+public:
+  PerThreadBoolIndicator(){};
+
+  BoolIndicatorUInt64& operator[]( const thread tid );
```
I would say: To be added as the need arises.

jakobj replied:
if they'd be provided, wouldn't such calls as in https://github.com/nest/nest-simulator/blob/4b15e1aef8f86e74b8d60ef9fc8e0391764a9297/nestkernel/connection_manager.cpp#L558 use the const version? depends on what you mean by "need" ;)

suku248 replied:
I think `check_primary_connections_` would have to be `const`  in order to result in a call to the additional constant `operator[]` function you are suggesting. But please feel free to try this out and, of course, adapt the code in case you are right. I just don't want to add a function that is currently not used.

suku248 commented on nestkernel/event_delivery_manager.cpp:
```diff
@@ -308,16 +307,17 @@ EventDeliveryManager::gather_spike_data_( const thread tid,
-    gather_completed_checker_.set( tid, true );
+// Assume this is the last gather round and change to false
+// otherwise
+#pragma omp barrier
```
Thanks for spotting this! I had added the barriers as a precaution and then forgot about them.
I have removed them now.

suku248 commented on nestkernel/event_delivery_manager.cpp:
```diff
@@ -583,19 +583,20 @@ EventDeliveryManager::gather_target_data( const thread tid )
-    gather_completed_checker_.set( tid, true );
+// assume this is the last gather round and change to false
+// otherwise
+#pragma omp barrier
```
See above.

suku248 commented on nestkernel/per_thread_bool_indicator.h:
```diff
@@ -0,0 +1,105 @@
+  static const uint_fast64_t false_uint64 = false;
+  uint_fast64_t status_;
+
+  friend class PerThreadBoolIndicator;
```
Good point! Thanks! This was required in an earlier version of the code. I have removed this now.

suku248 commented:
@jakobj @heplesser Thank you for your suggestions! I think I have addressed all of your comments. Please take another look.

I have repeatedly run the `hpc_benchmark` (adapted to use only a single synapses type for all outgoing connections of excitatory neurons, see #1394) using 32 MPI processes and 24 threads per process (at least 40 repetitions). With the code changes introduced in this PR I no longer observe any variations in spike counts across different runs.

jakobj commented on nestkernel/per_thread_bool_indicator.cpp:
```diff
@@ -0,0 +1,137 @@
+void
+BoolIndicatorUInt64::logical_and( const bool status )
+{
+  status_ = ( static_cast< bool >( status_ ) and status );
```
maybe something like this? but i don't feel too strongly about it

```suggestion
  if ( static_cast< bool >( status_ ) and status )
  {
      set_true();
  }
  else
  {
      set_false();
  }
```

diesmann replied:
I find the one line version more clear and the static_cast<> expresses that a sequence of conversions is going on.

suku248 replied:
I have not used an `if/else`-solution as I think the one-line version is unambiguous. It is well defined by the C++ International Standard what happens when assigning a `bool` value to an integer variable.

heplesser replied:
I agree with @suku248. The `if-else` is indeed to complex. If one wanted to be explicit, one could use
```c++
status_ = ( static_cast< bool >( status_ ) and status ) ? true_uint64 : false_uint64;
```
but I think that confuses more than it clarifies.
